### PR TITLE
feat: support enable/disable route

### LINKED ISSUE #2737: Proposal: Support enable/disable route
### Issue description

We could add enable/disable status to route so that it would support online / offline feature.

refer to:  https://github.com/apache/apisix-dashboard/issues/611#issuecomment-722256486

### PR DESCRIPTION
**Pull request by tzssangglass**
fix #2737

### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by liuxiran** [comment]
Thanks for you pr :), just wait for others preview.

**Comment by membphis** [comment]
nice PR

**Review by membphis — CHANGES_REQUESTED** [review]
please fix other similar points

**Comment by tzssangglass** [comment]
solve above

**Comment by tzssangglass** [comment]
describe a strange phenomenon, I was using `/route_status` api and create new routes because I found that using route id 1 and `/hello` always failed when I ran my local test push, even through other test cases, like

```
$prove -I../test-nginx/lib -I./ -r -s t/router/radixtree-host-uri.t         
t/router/radixtree-host-uri.t .. 10/? 
#   Failed test 'TEST 5: hit routes - status code ok'
#   at /usr/local/apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 948.
#          got: '404'
#     expected: '200'

#   Failed test 'TEST 5: hit routes - response_body - response is expected (repeated req 0, req 0)'
#   at /usr/local/apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 1589.
#          got: '{"error_msg":"404 Route Not Found"}
# '
#     expected: 'hello world
# '

#   Failed test 'TEST 6: hit routes - status code ok'
#   at /usr/local/apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 948.
#          got: '404'
#     expected: '200'

#   Failed test 'TEST 6: hit routes - response_body - response is expected (repeated req 0, req 0)'
#   at /usr/local/apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 1589.
#          got: '{"error_msg":"404 Route Not Found"}
# '
#     expected: 'hello world
# '
```
until I debugged at https://github.com/apache/apisix/blob/9e3c34a0103b88d340285017e34021cbc1f0fde8/apisix/http/router/radixtree_uri.lua#L85

when run test, the `user_routes.values` is nil, but when start apisix, it work.

after find this exception, something change, run the test case is back to normal, and cannot be reproduced. >_<|||

File: apisix/http/router/radixtree_host_uri.lua [hunk 1]
```diff
@@ -94,7 +94,14 @@ local function create_radixtree_router(routes)
     host_router = nil
 
     for _, route in ipairs(routes or {}) do
+        local status = core.table.try_read_attr(route, "value", "status")
+        if status and status ~= 1 then
```
Comment (tokers) [inline apisix/http/router/radixtree_host_uri.lua:98]: magic number is tricky, what about using literal string here?

Comment (tzssangglass) [reply apisix/http/router/radixtree_host_uri.lua:98]: just follow this
https://github.com/apache/apisix/blob/9e3c34a0103b88d340285017e34021cbc1f0fde8/apisix/schema_def.lua#L630-L635

File: apisix/schema_def.lua [hunk 2]
```diff
@@ -496,6 +496,13 @@ _M.route = {
         },
 
         id = id_schema,
+
+        status = {
+            description = "route status, 1 to enable, 0 to disable",
+            type = "integer",
+            enum = {1, 0},
```
Comment (tzssangglass) [inline apisix/schema_def.lua:503]: @tokers the `status` is define here

File: t/node/route-status.t [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+GET /hello
+--- error_code: 404
+--- response_body
+{"error_msg":"404 Route Not Found"}
+--- no_error_log
+[error]
+
+
+
+=== TEST 5: default enable route(id: 1)
```
Comment (liuxiran) [inline t/node/route-status.t:129]: it  is better to change the title to
`default enable route (id: 1) with host`

or someone may mistoke TEST 5 as TEST 1, e. g:me😂

**Comment by liuxiran** [comment]
is it necessary to add  a case：
create a route disabled -> hit route failed-> enable it -> hit route successfully?

**Comment by liuxiran** [comment]
> describe a strange phenomenon, I was using `/route_status` api and create new routes because I found that using route id 1 and `/hello` always failed when I ran my local test cases, even through other test cases, like
> 
> ```
> $prove -I../test-nginx/lib -I./ -r -s t/router/radixtree-host-uri.t         
> t/router/radixtree-host-uri.t .. 10/? 
> #   Failed test 'TEST 5: hit routes - status code ok'
> #   at /usr/local/apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 948.
> #          got: '404'
> #     expected: '200'
> 
> #   Failed test 'TEST 5: hit routes - response_body - response is expected (repeated req 0, req 0)'
> #   at /usr/local/apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 1589.
> #          got: '{"error_msg":"404 Route Not Found"}
> # '
> #     expected: 'hello world
> # '
> 
> #   Failed test 'TEST 6: hit routes - status code ok'
> #   at /usr/local/apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 948.
> #          got: '404'
> #     expected: '200'
> 
> #   Failed test 'TEST 6: hit routes - response_body - response is expected (repeated req 0, req 0)'
> #   at /usr/local/apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 1589.
> #          got: '{"error_msg":"404 Route Not Found"}
> # '
> #     expected: 'hello world
> # '
> ```
> 
> until I debugged at
> 
> https://github.com/apache/apisix/blob/9e3c34a0103b88d340285017e34021cbc1f0fde8/apisix/http/router/radixtree_uri.lua#L85
> 
> when run test, the `user_routes.values` is nil, but when start apisix, it work.
> 
> after find this exception, something change, run the test case is back to normal, and cannot be reproduced. >_<|||

If this magical event happened next, you may cc
 @membphis @spacewander  @tokers for help😊

File: t/node/route-status.t (hunk 3 above)
Comment (tzssangglass) [reply t/node/route-status.t:129]: good idea

File: apisix/http/router/radixtree_host_uri.lua (hunk 1 above)
Comment (membphis) [reply apisix/http/router/radixtree_host_uri.lua:98]: add comments?

File: apisix/http/router/radixtree_uri.lua [hunk 4]
```diff
@@ -38,6 +38,11 @@ local function create_radixtree_router(routes)
 
     for _, route in ipairs(routes) do
         if type(route) == "table" then
+            local status = core.table.try_read_attr(route, "value", "status")
+            if status and status ~= 1 then
```
Comment (membphis) [inline apisix/http/router/radixtree_uri.lua:42]: we can use `if status == 0 then` here, it seems clearer

File: t/node/route-status.t (hunk 3 above)
Comment (membphis) [reply t/node/route-status.t:129]: not update?

File: t/node/route-status.t [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 6: hit routes
```
Comment (membphis) [inline t/node/route-status.t:164]: Singular, `hit route`

**Comment by tzssangglass** [comment]
> is it necessary to add a case：
> create a route disabled -> hit route failed-> enable it -> hit route successfully?

test case 1 ~ 4 is testing for uri match, the order is 
create a route enable(default) -> hit route successfully-> disable it -> hit route fail

test case 5 ~ 8 is testing for host + uri match, the order is same as above.

I think combining these four steps into one test case would have the same effect.

**Review by membphis — APPROVED** [review]
LGTM, except minor style things

File: apisix/schema_def.lua (hunk 2 above)
Comment (membphis) [inline apisix/schema_def.lua:503]: add new test case: specify an invalid status value, eg: `100`

Comment (tzssangglass) [reply apisix/schema_def.lua:503]: add it

File: t/node/route-status.t (hunk 5 above)
Comment (tzssangglass) [reply t/node/route-status.t]: solve

File: apisix/http/router/radixtree_uri.lua (hunk 4 above)
Comment (tzssangglass) [reply apisix/http/router/radixtree_uri.lua]: solve

File: apisix/http/router/radixtree_host_uri.lua [hunk 6]
```diff
@@ -95,7 +95,8 @@ local function create_radixtree_router(routes)
 
     for _, route in ipairs(routes or {}) do
         local status = core.table.try_read_attr(route, "value", "status")
-        if status and status ~= 1 then
+        -- check the status
+        if status and status == 0 then
```
Comment (membphis) [inline apisix/http/router/radixtree_host_uri.lua:99]: only `if status == 0 then`, this is enough

Comment (tzssangglass) [reply apisix/http/router/radixtree_host_uri.lua:99]: will there be incompatibilities?
such as an existing route in etcd does not have the `status` attribute?

Comment (moonming) [reply apisix/http/router/radixtree_host_uri.lua:99]: You should add test cases for it:)

File: apisix/http/router/radixtree_host_uri.lua [hunk 7]
```diff
@@ -94,7 +94,15 @@ local function create_radixtree_router(routes)
     host_router = nil
 
     for _, route in ipairs(routes or {}) do
+        local status = core.table.try_read_attr(route, "value", "status")
+        -- check the status
+        if status and status == 0 then
+            goto CONTINUE
```
Comment (moonming) [inline apisix/http/router/radixtree_host_uri.lua:100]: Do we really need "goto"?

File: apisix/http/router/radixtree_host_uri.lua (hunk 6 above)
Comment (tzssangglass) [reply apisix/http/router/radixtree_host_uri.lua:99]: yes, i will

File: doc/admin-api.md [hunk 8]
```diff
@@ -72,6 +72,7 @@
 |service_protocol|False|Upstream protocol type|only `grpc` and `http` are supported|`http` is the default value; Must set `grpc` if using `gRPC proxy` or `gRPC transcode`|
 |labels   |False |Match Rules|Key/value pairs to specify attributes|{"version":"v2","build":"16","env":"production"}|
 |enable_websocket|False|Auxiliary| enable `websocket`(boolean), default `false`.||
+|status          |False|Auxiliary| enable this route, default `1`.|`1` to enable, `0` to disable|
```
Comment (moonming) [inline doc/admin-api.md:75]: Need to add examples for how to use this.

File: doc/zh-cn/admin-api.md [hunk 9]
```diff
@@ -76,6 +76,7 @@
 |filter_func|可选|匹配规则|用户自定义的过滤函数。可以使用它来实现特殊场景的匹配要求实现。该函数默认接受一个名为 vars 的输入参数，可以用它来获取 Nginx 变量。|function(vars) return vars["arg_name"] == "json" end|
 |labels   |可选 |匹配规则|标识附加属性的键值对|{"version":"v2","build":"16","env":"production"}|
 |enable_websocket|可选 |辅助| 是否启用 `websocket`(boolean), 缺省 `false`.||
+|status          |可选 |辅助| 是否启用此路由, 缺省 `1`。|`1` 表示启用，`0` 表示禁用|
 |create_time|可选|辅助|单位为秒的 epoch 时间戳，如果不指定则自动创建|1602883670|
```
Comment (moonming) [inline doc/zh-cn/admin-api.md:79]: Why the default time is not "now"? @nic-chen @membphis

File: apisix/http/router/radixtree_host_uri.lua (hunk 7 above)
Comment (tzssangglass) [reply apisix/http/router/radixtree_host_uri.lua:100]: you mean like this?
```
        if status = 1 then
              push_host_router(route, host_routes, only_uri_routes)
        end
```

File: doc/admin-api.md (hunk 8 above)
Comment (tzssangglass) [reply doc/admin-api.md:75]: get

File: apisix/http/router/radixtree_host_uri.lua (hunk 7 above)
Comment (moonming) [reply apisix/http/router/radixtree_host_uri.lua:100]: yes

Comment (moonming) [reply apisix/http/router/radixtree_host_uri.lua:100]: and we need to deal with old data which not has status

Comment (tzssangglass) [reply apisix/http/router/radixtree_host_uri.lua:100]: get

File: doc/zh-cn/admin-api.md (hunk 9 above)
Comment (nic-chen) [reply doc/zh-cn/admin-api.md:79]: It should be a typo here.  @spacewander  please have a look.

Comment (spacewander) [reply doc/zh-cn/admin-api.md:79]: @moonming @nic-chen 
The last item `1602883670` is an example, not the default value.

**Comment by moonming** [comment]
got it

Thanks,
Ming Wen
Twitter: _WenMing


罗泽轩 <notifications@github.com> 于2020年12月5日周六 下午9:25写道：

File: apisix/http/router/radixtree_host_uri.lua [hunk 10]
```diff
@@ -94,7 +94,11 @@ local function create_radixtree_router(routes)
     host_router = nil
 
     for _, route in ipairs(routes or {}) do
-        push_host_router(route, host_routes, only_uri_routes)
+        local status = core.table.try_read_attr(route, "value", "status")
+        -- check the status
+        if status == 1 then
```
Comment (membphis) [inline apisix/http/router/radixtree_host_uri.lua:99]: The existing old data has no field "status". 

we need to make sure it works fine for this case.

`if not status or status == 1 then`, this way is safer.

Comment (tzssangglass) [reply apisix/http/router/radixtree_host_uri.lua:99]: I took a look at the apisix sync data from etcd,
https://github.com/apache/apisix/blob/c41aa8534a18f43ef0c3cd6780f106595f9f9b88/apisix/core/schema.lua#L53-L61

after `validator(json) `, the existing old data will has default `status` value. 

but i agree your point, add `if not status` is more safer:-)

File: t/node/route-status.t [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+GET /t
+--- error_code: 400
+--- response_body eval
+qr/\{"error_msg":"invalid configuration: property \\"status\\" validation failed: matches non of the enum values"\}/
+--- no_error_log
+[error]
+
+
+
+=== TEST 10: compatible with old rpote data in etcd which not has status
```
Comment (Firstsawyou) [inline t/node/route-status.t:264]: `rpote` Is this a typo, it should be `route`, right?

Comment (tzssangglass) [reply t/node/route-status.t:264]: > `rpote` Is this a typo, it should be `route`, right?

yep

**Review by spacewander — APPROVED** [review]

**Comment by tzssangglass** [comment]
thanks to the people at code review, and thanks for the guidance!

**Pull request merged**